llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -1484,6 +1484,11 @@ bool RISCVTargetLowering::shouldExpandGetVectorLength(EVT TripCountVT,
   return VF > MaxVF || !isPowerOf2_32(VF);
 }
 
+bool RISCVTargetLowering::shouldExpandCttzElements(EVT VT) const {
+  return !Subtarget.hasVInstructions() ||
+         VT.getVectorElementType() != MVT::i1 || !isTypeLegal(VT);
```

Comment:
We could handle the legal non-i1 case via s simple vector comparison against zero - producing a mask result that then flows through this code.  From the tests, this looks like a pretty huge improvement in codegen quality.